src/Dapr.Messaging/PublishSubscribe/DaprPublishSubscribeClient.cs

```diff
@@ -0,0 +1,64 @@
+﻿// ------------------------------------------------------------------------
+// Copyright 2024 The Dapr Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//     http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ------------------------------------------------------------------------
+
+namespace Dapr.Messaging.PublishSubscribe;
+
+/// <summary>
+/// The base implementation of a Dapr pub/sub client.
+/// </summary>
+public abstract class DaprPublishSubscribeClient
+{
+    /// <summary>
+    /// Dynamically subscribes to a Publish/Subscribe component and topic.
+    /// </summary>
+    /// <param name="pubsubName">The name of the Publish/Subscribe component.</param>
+    /// <param name="topicName">The name of the topic to subscribe to.</param>
+    /// <param name="options">Configuration options.</param>
+    /// <param name="cancellationToken">Cancellation token.</param>
+    /// <returns>An <see cref="IAsyncEnumerable{TopicMessage}"/> containing the various messages returned by the subscription.</returns>
+    public abstract IAsyncEnumerable<TopicMessage> SubscribeAsync(string pubsubName, string topicName, DaprSubscriptionOptions options, CancellationToken cancellationToken);
```

Comment:
Are you subscribed to the messages after completion of `SubscribeAsync()` or only after calling `GetAsyncEnumerator()` on the resulting `IAsyncEnumerator`?  If the former, are messages buffered until enumeration actually starts?  If the latter, can multiple calls to `GetAsyncEnumerator()` be made?

Reply:
(I think I have my answers having looked at the implementation, but we may want to be explicit about the semantics here so that the semantics cannot vary at the whims of the current implementation.)

Reply:
I've switched this up to more closely resemble the implementation you started with. I agree that it's just cleaner and far more simply accommodates the sequential acknowledgement (didn't realize that initially) without having all these other methods the developer is responsible for calling here and there. The handler just sets up callers up for success without having to think much about what's happening behind the scenes.

As of the latest commit, the user calls Register on the `DaprPublishSubscribeGrpcClient` and that returns the `PublishSubscribeReceiver` corresponding to the pubsub component name and topic name (or throws if that combination is already registered). The developer then calls `SubscribeAsync` on that and it kicks off an unbounded operation to read the messages from the sidecar to a Channel on the type (for backpressure purposes). In the same method, the messages are read out of the Channel one by one and a `CancellationTokenSource` linked with the timeout value specified in the configuration options. Each message is evaluated by the handler delegate which returns an action to take and either this result is passed back to the sidecar or the cancellation token will have timed out and the default action returned.

This way, the messages are buffered if they're returned faster than the handler is evaluating them at any given moment. As a mention in another comment in here somewhere, this leaves the door open for this implementation to eventually support consumer groups/partitions at some later point to allow richer back-pressure mitigation.